sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,224 @@
+        The metric to use when calculating distance between instances in a
+        feature array. If metric is a string or callable, it must be one of
+        the options allowed by metrics.pairwise.calculate_distance for its
+        metric parameter.
+        If metric is "precomputed", X is assumed to be a distance matrix and
+        must be square.
+
+    extraction : string, optional
+        The extraction method used to generate clusters from the ordering of
+        points returned by the OPTICS algorithm.
+
+    ext_kwargs : dict
+        Keyword arguments to be supplied to the extraction function.
+
+    Returns
+    -------
+    core_distances : array [n_samples]
+        Core distance for each sample.
+
+    ordering : array [n_samples]
+        Indices of the samples in the order generated by OPTICS.
+
+    reachability_distances : array [n_samples]
+        Reachability distance for each sample.
+
+    labels : array [n_samples]
+        Cluster labels for each point. Noisy samples are given the label -1.
+
+    Notes
+    -----
+    See examples/cluster/plot_optics.py for an example.
+
+    References
+    ----------
+    Ankerst, Mihael, Markus M. Breunig, Hans-Peter Kriegel, and Jörg Sander.
+    "OPTICS: ordering points to identify the clustering structure." ACM SIGMOD
+    Record 28, no. 2 (1999): 49-60.
+
+    """
+    X = atleast2d_or_csr(X)
+    n = X.shape[0]
+    if min_samples > n:
+        raise ValueError('min_samples must be lower than the total number of samples')
+    ordering = []
+    core_distances = np.ndarray(len(X))
+    # Initiate reachability distances to infinity
+    reachability_distances = float('inf') * np.ones(n)
+    # Set reachability for first point
+    reachability_distances[0] = 0
+    # Construct spatial indexing structure
+    if metric != 'precomputed':
+        # TODO: Construct BallTree with the correct metric once the
+        # metrics branch has been merged into master
+        tree = BallTree(X, metric=metric)
+
+    seeds = np.ones(n, dtype=bool)
+    i = 0
+    while True:
+        # Mark current point as processed
+        seeds[i] = False
+        # Add current point to the ordering
+        ordering.append(i)
+        if not any(seeds):
+            break
+        # Calculate core distance
+        if metric == 'precomputed':
+            D = X[i]
+            core_dist = np.sort(D)[min_samples]
+        else:
+            core_dist = tree.query(X[i], min_samples+1)[0][0][-1]
+        core_distances[i] = core_dist
+
+        if core_dist <= eps:
+            # Get the neighbors of the current point
+            if metric == 'precomputed':
+                neighbors = D[seeds] <= eps
+                ds = D[neighbors]
+            else:
+                ind, dist = tree.query_radius(X[i], eps, True)
+                si = seeds[ind[0]]
+                neighbors = ind[0][si]
+                ds = dist[0][si]
+            cds = core_dist * np.ones(len(ds))
```

Comment:
This is unnecessary. `np.maximum(core_dist, ds)` should suffice due to broadcasting rules.